test/gsb-test-srv/main.go

```diff
@@ -0,0 +1,468 @@
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	gsb "github.com/letsencrypt/boulder/test/gsb-test-srv/proto"
+)
+
+// testSrv implements a bare bones mock Google Safe Browsing server. The `hits`
+// and `hp` fields are protected against concurrent updates using `mu`.
+type testSrv struct {
+	apiKey string
+	list   safebrowsingList
+	hits   map[string]int
+	mu     *sync.RWMutex
+}
+
+const (
+	protoMime = "application/x-protobuf"
+)
+
+// defaultUnsafeURLs is the list of URLs that we return a "unsafe" response for.
+// Additional URLs are added by sending POSTs to the `/add` HTTP endpoint.
+var defaultUnsafeURLs = []string{
+	"honest.achmeds.discount.hosting.com",
+}
+
+// emptyThreatListUpdateResp is an empty threat list update response for padding
+// responses out to include the correct number of list updates clients expect
+var emptyThreatListUpdateResp = &gsb.FetchThreatListUpdatesResponse_ListUpdateResponse{
+	ThreatType:      gsb.ThreatType_MALWARE,
+	PlatformType:    gsb.PlatformType_ANY_PLATFORM,
+	ThreatEntryType: gsb.ThreatEntryType_URL,
+	ResponseType:    gsb.FetchThreatListUpdatesResponse_ListUpdateResponse_FULL_UPDATE,
+	/*
+	 * This is the SHA1Sum of `[]byte{}`, e.g. of an empty list of additions
+	 */
+	Checksum: &gsb.Checksum{
+		Sha256: []byte{
+			0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4, 0xc8,
+			0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c,
+			0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55,
+		},
+	},
+}
+
+// unmarshalPB unmarshals a request body into a protocol buffer message
+func unmarshalPB(req *http.Request, pbReq proto.Message) error {
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return err
+	}
+	if err := proto.Unmarshal(body, pbReq); err != nil {
+		return err
+	}
+	return nil
+}
+
+// marshalPB marshals a protocol buffer message into a response body
+func marshalPB(resp http.ResponseWriter, pbResp proto.Message) error {
+	resp.Header().Set("Content-Type", protoMime)
+	body, err := proto.Marshal(pbResp)
+	if err != nil {
+		return err
+	}
+	if _, err := resp.Write(body); err != nil {
+		return err
+	}
+	return nil
+}
+
+// listEntry represents an entry on the safe browsing list. Each entry is a hash
+// constructed out of the URL.
+type listEntry struct {
+	hash string
```

Comment:
Update: turns out you can't compare two fixed size arrays. Consensus is to leave this as-is.